### PR TITLE
Switch CLI to async

### PR DESCRIPTION
**Pull request by fisker**
## Description

This PR didn't actually use any async API in CLI, just make it async and make test pass.
This is a necessary step for #4459 and #4980

Finally, I figured how to test it.

~There is still [one failed test](https://github.com/prettier/prettier/pull/10804/files#diff-d30da97d630a4c42ca81fcfd745d6b2ac8f4f757e1e7c786e9189bc0a7110f21R90), I can't debug it on windows, maybe @thorn0 or @connor4312 can help me out?~ Fixed

## Checklist

- [ ] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [ ] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by fisker** [comment]
Not ready yet.

**Comment by connor4312** [comment]
Just moving the creation of directories into `beforeAll` fixes it for me. Guessing these may have been clobbered by concurrently-running tests or something.

<details><summary>patch</summary>

```diff
diff --git a/tests/integration/__tests__/patterns-dirs.js b/tests/integration/__tests__/patterns-dirs.js
index 39f7f986f..27d129566 100644
--- a/tests/integration/__tests__/patterns-dirs.js
+++ b/tests/integration/__tests__/patterns-dirs.js
@@ -1,7 +1,7 @@
 "use strict";
 
 const path = require("path");
-// const fs = require("fs");
+const fs = require("fs");
 const runPrettier = require("../runPrettier");
 const { projectRoot } = require("../env");
 
@@ -87,38 +87,37 @@ describe("plugins `*`", () => {
   });
 });
 
-// TODO: fix this test
-// if (path.sep === "/") {
-//   // Don't use snapshots in these tests as they're conditionally executed on non-Windows only.
-
-//   const base = path.resolve(__dirname, "../cli/patterns-dirs");
-
-//   // We can't commit these dirs without causing problems on Windows.
-
-//   // TODO: these should be moved to a `beforeAll`, but for that to be possible,
-//   // `runPrettier` should be refactored to use `describe` and `beforeEach` for doing setup.
-//   fs.mkdirSync(path.resolve(base, "test-a\\"));
-//   fs.writeFileSync(path.resolve(base, "test-a\\", "test.js"), "x");
-//   fs.mkdirSync(path.resolve(base, "test-b\\?"));
-//   fs.writeFileSync(path.resolve(base, "test-b\\?", "test.js"), "x");
-
-//   describe("Backslashes in names", () => {
-//     afterAll(() => {
-//       fs.unlinkSync(path.resolve(base, "test-a\\", "test.js"));
-//       fs.rmdirSync(path.resolve(base, "test-a\\"));
-//       fs.unlinkSync(path.resolve(base, "test-b\\?", "test.js"));
-//       fs.rmdirSync(path.resolve(base, "test-b\\?"));
-//     });
-
-//     testPatterns("", ["test-a\\/test.js"], { stdout: "test-a\\/test.js\n" });
-//     testPatterns("", ["test-a\\"], { stdout: "test-a\\/test.js\n" });
-//     testPatterns("", ["test-a*/*"], { stdout: "test-a\\/test.js\n" });
-
-//     testPatterns("", ["test-b\\?/test.js"], { stdout: "test-b\\?/test.js\n" });
-//     testPatterns("", ["test-b\\?"], { stdout: "test-b\\?/test.js\n" });
-//     testPatterns("", ["test-b*/*"], { stdout: "test-b\\?/test.js\n" });
-//   });
-// }
+if (path.sep === "/") {
+  // Don't use snapshots in these tests as they're conditionally executed on non-Windows only.
+
+  const base = path.resolve(__dirname, "../cli/patterns-dirs");
+
+  // We can't commit these dirs without causing problems on Windows.
+
+  describe("Backslashes in names", () => {
+    beforeAll(() => {
+      fs.mkdirSync(path.resolve(base, "test-a\\"));
+      fs.writeFileSync(path.resolve(base, "test-a\\", "test.js"), "x");
+      fs.mkdirSync(path.resolve(base, "test-b\\?"));
+      fs.writeFileSync(path.resolve(base, "test-b\\?", "test.js"), "x");
+    });
+
+    afterAll(() => {
+      fs.unlinkSync(path.resolve(base, "test-a\\", "test.js"));
+      fs.rmdirSync(path.resolve(base, "test-a\\"));
+      fs.unlinkSync(path.resolve(base, "test-b\\?", "test.js"));
+      fs.rmdirSync(path.resolve(base, "test-b\\?"));
+    });
+
+    testPatterns("", ["test-a\\/test.js"], { stdout: "test-a\\/test.js\n" });
+    testPatterns("", ["test-a\\"], { stdout: "test-a\\/test.js\n" });
+    testPatterns("", ["test-a*/*"], { stdout: "test-a\\/test.js\n" });
+
+    testPatterns("", ["test-b\\?/test.js"], { stdout: "test-b\\?/test.js\n" });
+    testPatterns("", ["test-b\\?"], { stdout: "test-b\\?/test.js\n" });
+    testPatterns("", ["test-b*/*"], { stdout: "test-b\\?/test.js\n" });
+  });
+}
 
 function testPatterns(namePrefix, cliArgs, expected = {}) {
   const testName =
```

</details>

**Comment by fisker** [comment]
@connor4312 Thank you! I'll update when I get home.

**Comment by connor4312** [comment]
🚀

**Comment by fisker** [comment]
Maybe, but no promise, a little busy recently.

**Comment by connor4312** [comment]
Ok. If I get some time this weekend, I'll start on that. Maybe it would be easiest to collaborate on a single branch to avoid too much conflict. Do you want to make a fork/branch of prettier that I can write to? Or I can do so.

File: tests/integration/runPrettier.js [hunk 1]
```diff
@@ -113,11 +113,60 @@ function runPrettier(dir, args = [], options = {}) {
...
+    if (hasRunningCli) {
+      throw new Error("Please wait for previous CLI to exit.");
+    }
+
+    if (!promise) {
+      hasRunningCli = true;
+      promise = run(dir, args, options);
+      promise.then(() => {
+        hasRunningCli = false;
+      });
```
Comment (thorn0) [inline tests/integration/runPrettier.js:161]:
```suggestion
      promise = run(dir, args, options).finally(() => {
        hasRunningCli = false;
      });
```
This way it's simpler to understand when exactly `hasRunningCli = false;` is executed.

**Pull request merged**